### PR TITLE
Dont introduce aliasing in CSE or Constant Pooling

### PR DESCRIPTION
**Pull request by eellison**
We can't introduce aliasing to a graph output, since they may be mutated after.

### CODE REVIEW

File: aten/src/ATen/core/jit_type.h [hunk 1]
```diff
@@ -1061,7 +1061,7 @@ struct CAFFE2_API StringType : public Type {
     return rhs.kind() == kind();
   }
   std::string str() const override {
-    return "string";
+    return "str";
```
Comment (suo) [inline aten/src/ATen/core/jit_type.h]: This change seems unrelated. Can you split it into a different PR?

File: torch/csrc/jit/passes/alias_analysis.h [hunk 2]
```diff
@@ -49,6 +49,23 @@ class AliasDb {
...
   bool mayContainAlias(const Value* a, const Value* b) const;
 
+  // Do any values in group `a` share a memory location or hold in memory
+  // any element that exists in group `b`
+  template <
+      typename... Other1,
+      template <typename, typename...> class T,
+      typename... Other2,
+      template <typename, typename...> class U>
+  bool mayContainAlias(
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.h]: I feel like we keep re-introducing this inefficient pairwise comparison pattern in different levels of the code. Avoiding that is the reason `mayAlias` is implemented the way it is below, where we gather all memory locations before comparing them for overlap. Let's do the same here.

**Review by suo — CHANGES_REQUESTED** [review]

File: aten/src/ATen/core/jit_type.h (hunk 1 above)
Comment (eellison) [reply aten/src/ATen/core/jit_type.h]: It was a necessary fix to make the test in constant pooling work with this change. Previously "string" was being outputted and parsed as a typevar.

File: torch/csrc/jit/passes/alias_analysis.h (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.h]: I would rather avoid the way it's done below since it effectively implements mayAlias in two different locations. E.g. below this there is no check for 
`if (!shouldAnnotate(a) || !shouldAnnotate(b)) {` but that has since been added to the mayAlias implementation in the cpp file. 

If we do want to avoid gathering all memory locations, i think it makes the most sense to add an API in memory_dag.cpp for the group checking

Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.h]: Good point, but that's a reason to delete the singleton variant of `mayAlias` I think (or at least make it a proxy for the container version). In general, we should be wary of code patterns that involve a lot of unnecessary work in passes/methods that are called really frequently.

I agree that we'll probably need to add to memory_dag (similar to how mayAlias has a memory_dag container comparison implementation as well).

File: test/test_jit.py [hunk 3]
```diff
@@ -12703,28 +12720,6 @@ def fn(x, k):
         # the same group; they should each be a separate DiffGraph
         self.assertGraphContainsExactly(graph, 'prim::DifferentiableGraph', 2)
 
-    def test_mutation_subgraph_inlining(self):
```
Comment (ZolotukhinM) [inline test/test_jit.py]: Why did this test disappear? :)

Comment (eellison) [reply test/test_jit.py]: See the accompanied change in create_autodiff_subgraphs.cpp.

File: torch/csrc/jit/passes/alias_analysis.h [hunk 4]
```diff
@@ -47,7 +47,39 @@ class AliasDb {
 
   // Does `a` and `b` potentially share a memory location or do either
   // hold in memory any element that exists in the other
-  bool mayContainAlias(const Value* a, const Value* b) const;
+  bool mayContainAlias(Value* a, Value* b) const {
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.h]: nit: put these implementations in the cpp file, since they are not templates

**Review by suo — APPROVED** [review]
lgtm, small nit inline
